diff -r 71a233cd4e5d pySvnManager.egg-info/paste_deploy_config.ini_tmpl
--- a/pySvnManager.egg-info/paste_deploy_config.ini_tmpl	Sat Jul 04 16:54:01 2009 +0800
+++ b/pySvnManager.egg-info/paste_deploy_config.ini_tmpl	Sun Jul 05 10:17:58 2009 +0800
@@ -14,7 +14,7 @@
 host = 0.0.0.0
 port = 5000
 
-[app:main]
+[app:myapp]
 use = egg:pySvnManager
 full_stack = true
 cache_dir = %(here)s/data
@@ -50,11 +50,14 @@
 authz_file = %(here)s/config/svn.access
 
 # svn repository root path
-repos_root = %(here)s/svnroot
+repos_root = /opt/svn/svnroot
 
-#[filter:proxy-prefix]
-#use = egg:PasteDeploy#prefix
-#prefix = /svnadmin
+[filter:proxy-prefix]
+use = egg:PasteDeploy#prefix
+prefix = /svnadmin
+
+[pipeline:main]
+pipeline = proxy-prefix myapp
 
 # Logging configuration
 [loggers]
